Booktour/src/main/java/app/dao/GenericDAO.java

```diff
@@ -0,0 +1,53 @@
+package app.dao;
+
+import java.io.Serializable;
+import java.lang.reflect.ParameterizedType;
+
+import org.hibernate.Session;
+import org.hibernate.SessionFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.orm.hibernate5.support.HibernateDaoSupport;
+
+public abstract class GenericDAO<PK extends Serializable, T> extends HibernateDaoSupport {
+
+	@Autowired
+	private SessionFactory sessionFactory;
+
+	public T findById(Serializable key) {
+		return (T) getSession().get(getPersistentClass(), key);
+	}
+
+	public void persist(T entity) {
+		getSession().persist(entity);
+	}
+
+	public void delete(T entity) {
+		getSession().delete(entity);
+	}
+
+	public T saveOrUpdate(T entity) {
+		getSession().saveOrUpdate(entity);
+		return entity;
+	}
+
+	private Class<T> persistentClass;
+
+	public Class<T> getPersistentClass() {
+		return persistentClass;
+	}
+
+	public GenericDAO(Class<T> persistentClass) {
+		this.persistentClass = persistentClass;
+	}
+
+	@SuppressWarnings("unchecked")
+	public GenericDAO() {
+		this.persistentClass = (Class<T>) ((ParameterizedType) this.getClass().getGenericSuperclass())
```

Comment:
giải thích cho chị nghe code chỗ này xem nào

Reply:
em hiểu như thế này ạ. hàm này là  đóng vai trò như contructor ko tham số. nhận mọi loại type như object, Interger, String... 
 xét cả toàn abstract class này thì nó sẽ như class dùng chung cho toàn project mà chỉ cần gọi đến để dùng. đúng với OOP